File: src/main/java/AoC2017_06.java
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class AoC2017_06 extends SolutionBase<List<Integer>, Integer, Integer> {

    private AoC2017_06(final boolean debug) {
        super(debug);
    }

    public static AoC2017_06 create() {
        return new AoC2017_06(false);
    }

    public static AoC2017_06 createDebug() {
        return new AoC2017_06(true);
    }

    @Override
    protected List<Integer> parseInput(final List<String> inputs) {
        return Arrays.stream(inputs.getFirst().split("\\s+")).map(Integer::valueOf).toList();
    }

    private Result getRedistributionCycles(final List<Integer> input) {
        final Map<List<Integer>, Integer> seen = new HashMap<>();
        List<Integer> banks = new ArrayList<>(input);
        int cnt = 0;
        while (!seen.containsKey(banks)) {
            seen.put(banks, cnt);
            final int max = banks.stream().mapToInt(Integer::valueOf).max().getAsInt();
            final int idx = banks.indexOf(max);
            final Integer blocks = banks.get(idx);
            banks = new ArrayList<>(banks);
            banks.set(idx, 0);
            for (int i = 1; i <= blocks; i++) {
                final int j = (idx + i) % banks.size();
                final int newBlocks = banks.get(j) + 1;
                banks.set(j, newBlocks);
            }
            cnt++;
        }
        return new Result(seen, banks);
    }

    @Override
    public Integer solvePart1(final List<Integer> input) {
        return getRedistributionCycles(input).map.size();
    }

    @Override
    public Integer solvePart2(final List<Integer> input) {
        final Result result = getRedistributionCycles(input);
        return result.map.size() - result.map.get(result.last);
    }

    @Samples({
        @Sample(method = "part1", input = "0 2 7 0", expected = "5"),
        @Sample(method = "part2", input = "0 2 7 0", expected = "4"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }

    record Result(Map<List<Integer>, Integer> map, List<Integer> last) {}
}

File: src/main/python/AoC2017_06.py
#! /usr/bin/env python3
#
# Advent of Code 2017 Day 6
#
from aoc import my_aocd
import aocd


def _parse(inputs: tuple[str]) -> list[int]:
    return [int(_) for _ in inputs[0].split()]


def _solve(inputs: tuple[str]):
    banks = tuple(_parse(inputs))
    cnt = 0
    seen = dict()
    while banks not in seen:
        seen[banks] = cnt
        idx = banks.index(max(banks))
        blocks = banks[idx]
        _banks = [_ for _ in banks]
        _banks[idx] = 0
        for i in range(1, blocks + 1, 1):
            j = (idx + i) % len(_banks)
            _banks[j] += 1
        banks = tuple(_banks)
        cnt += 1
    return seen, banks


def part_1(inputs: tuple[str]) -> int:
    seen, _ = _solve(inputs)
    return len(seen)


def part_2(inputs: tuple[str]) -> int:
    seen, last = _solve(inputs)
    return len(seen) - seen[last]


TEST = """0 2 7 0""".splitlines()


def main() -> None:
    puzzle = aocd.models.Puzzle(2017, 6)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert part_1(TEST) == 5
    assert part_2(TEST) == 4

    inputs = my_aocd.get_input(2017, 6, 1)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == '__main__':
    main()
